### PR TITLE
[stdlib] Uses of static int tuple

### PR DESCRIPTION
**Pull request by artemiogr97**
Implementation for #2515

### CODE REVIEW

File: stdlib/src/builtin/int.mojo
Comment (gabrieldemarmiesse) [inline stdlib/src/builtin/int.mojo:700]:
```suggestion
            return div - 1, mod + rhs
```
This should be enough. By default, values separated by commas means that it's a tuple. Same for the other returns

File: stdlib/src/builtin/simd.mojo
Comment (gabrieldemarmiesse) [inline stdlib/src/builtin/simd.mojo:36]: **thoughts** I supposed this is here otherwise the compiler will complain that this does not exist. This is not the first time that I see this bug about the compiler not importing the builtins automatically. But it's really hard to reproduce.

**Review by gabrieldemarmiesse — COMMENTED** [review]
Thanks a lot for the pull request! That's a nice cleanup! Just a small simplification I found and it should be ok to merge this

File: stdlib/src/builtin/string.mojo
Comment (soraros) [inline stdlib/src/builtin/string.mojo:928]: I think we should remove this after switching its call sites to use `Tuple`. Maybe just leave a TODO here.

File: stdlib/src/builtin/simd.mojo
Comment (artemiogr97) [reply stdlib/src/builtin/simd.mojo:36]: Is `StaticIntTuple`  supposed to be a builtin? it is defined in utils 
As I mentionned here it was exposed as a builtin from the imports in the int module: https://github.com/modularml/mojo/issues/2515#issuecomment-2094760396

Comment (gabrieldemarmiesse) [reply stdlib/src/builtin/simd.mojo:36]: Then it seems you are importing it but not using it in simd.mojo. But maybe the PR is incomplete

Comment (artemiogr97) [reply stdlib/src/builtin/simd.mojo:36]: it was already used in simd.mojo, so I had to add the import here (it was missing but working "thanks" to the import in int.mojo)

Comment (gabrieldemarmiesse) [reply stdlib/src/builtin/simd.mojo:36]: Yeah that's really not a good thing how builtin imports work right now. I'll open an issue.

File: stdlib/src/builtin/string.mojo
Comment (artemiogr97) [reply stdlib/src/builtin/string.mojo:928]: I just tried removing it, and it works, should it remove it now?

Also I just realized that `String.join` does not match the python implementation of `str.join`, in python only one argument is expected, not a sequence of arguments

File: stdlib/src/builtin/simd.mojo
Comment (artemiogr97) [inline stdlib/src/builtin/simd.mojo:1623]: I didn't realize that I deleted this check, but I cannot get it to work, keeping ` <= mask[i] < 2 * size` gives the following error: 
```
no matching method in call to '__refitem__'mojo
static_tuple.mojo(341, 8): candidate not viable: callee expects at least 2 positional parameters, but 0 were specified
static_tuple.mojo(372, 8): candidate not viable: expected at most 1 positional argument, got 2
```

Comment (artemiogr97) [reply stdlib/src/builtin/simd.mojo:1623]: I saw that the reference changed a lot in the last nightly release so after updating the error changed to `cannot use a dynamic value in call parameter` which makes a lot more sense, but I'm still not sure how to add the compile time check again

**Comment by gabrieldemarmiesse** [comment]
For your information, https://github.com/modularml/mojo/pull/2097 was just merged internally, which means that you'll have conflicts in the `_divmod` method. You might want to wait for the next nightly and then rebase your changes. Between this and the `Reference` changes, I expect you'll have a fair amount of conflicts. If you want, then,  I can lend a hand :)

**Comment by artemiogr97** [comment]
@gabrieldemarmiesse I solved the conflicts for the `Reference` changes but I don't think I will have time to do much more these days, if you want I can give you write access to my fork and feel free to do a merge, otherwise I'll do it in the weekend.

Personally I'm a little bit more worried about the issue I mentioned above, I think that the check was important and maybe I'll have to revert the changes in simd.mojo to keep it.

**Comment by artemiogr97** [comment]
yep, and it's because of those holidays that I won't be home these days, i just sent you the invite 😉

**Comment by gabrieldemarmiesse** [comment]
Happy holiday :)

File: stdlib/src/builtin/simd.mojo
Comment (gabrieldemarmiesse) [reply stdlib/src/builtin/simd.mojo:1658]: Ok so I dig this one a bit and it's a compiler bug. The good news is that I found a workaround, as the struct `Tuple` had the same problem (I was wondering how Tuple handled this). So I found this: https://github.com/modularml/mojo/blob/nightly/stdlib/src/builtin/tuple.mojo#L182 I used the same trick here and it worked. It seems the issue is tracked internally, so this problem should go away at some point.

**Comment by gabrieldemarmiesse** [comment]
Hey @artemiogr97 , I had some time to kill:
* I fixed the conflicts, it's now all find for `_divmod`, which doesn't require changes in nightly
* Put back the compile-time check where you had some issues. See https://github.com/modularml/mojo/pull/2525#discussion_r1595590187
This PR should be ready for review :)

File: stdlib/src/builtin/string.mojo
Comment (JoeLoser) [reply stdlib/src/builtin/string.mojo:932]: Let's save the refactoring of `join` for a different PR, please.

File: stdlib/src/utils/static_tuple.mojo
Comment (JoeLoser) [inline stdlib/src/utils/static_tuple.mojo:396]: **Note to self** file this bug externally so we can track it easier

**Review by JoeLoser — APPROVED** [review]
Great cleanup! Thank you for this!

**Comment by JoeLoser** [comment]
> > @gabrieldemarmiesse I think @JoeLoser was talking about removing the method later, the branch already had the tag 'imported internally' see: [#2525 (comment)](https://github.com/modularml/mojo/pull/2525#discussion_r1590313416)
> 
> No worries, it's easy to update the internal PR. I'll re-import this. Thanks for the ping. We're working on internal tooling to notify when external changes have been made from imported internal PRs.
> 
> This is breaking some internal things relying on the specific IR of these stdlib types, but we're digging into it. Nothing needed from your part at all. 😄

I marked this PR as "blocked", but there's nothing for you to do.  There's some compiler issues that need to get sorted out internally before we can land this change, unfortunately.

**Comment by artemiogr97** [comment]
@JoeLoser I see that there are some conflicts now with this PR in several files, maybe it would be a good idea to split it?, also that way the "safe"/working bits could be merged.

**Comment by JoeLoser** [comment]
> @JoeLoser I see that there are some conflicts now with this PR in several files, maybe it would be a good idea to split it?, also that way the "safe"/working bits could be merged.

I'll take care of this, no worries — thank you for flagging this!  I want to split up/land as many as I can internally that aren't crashing the compiler.  It'll be easier for me to do this in your internal PR.

I'll take a look at this in about an hour from now.

**Comment by JoeLoser** [comment]
@artemiogr97 I tracked down the crash a bit, and it's due to the changes in `simd.mojo`.  I haven't root-caused it yet, but I will when I get some more time this week.  I backed off those changes (and the introduction of `get` function in `static_tuple.mojo` as it was only used in `simd.mojo` in this PR.  This should unblock this PR from landing in some shape at least here very shortly assuming internal CI passes after I reworked this PR.

**Pull request closed without merging**
